File: Heaps/PythonMinHeap/min_heap.py
class MinHeap:

    def __init__(self, capacity):
        self.storage = [0] * capacity
        self.capacity = capacity
        self.size = 0

    def getParentIdx(self, index):
        return (index - 1) // 2

    def getLeftChildIdx(self, index):
        return 2 * index + 1

    def getRightChildIdx(self, index):
        return 2 * index + 2

    def hasParent(self, index):
        return self.getParentIdx(index) >= 0

    def hasLeftChild(self, index):
        return self.getLeftChildIdx(index) < self.size

    def hasRightChild(self, index):
        return self.getRightChildIdx(index) < self.size

    def getParent(self, index):
        return self.storage[self.getParentIdx(index)]

    def getLeftChild(self, index):
        return self.storage[self.getLeftChildIdx(index)]

    def getRightChild(self, index):
        return self.storage[self.getRightChildIdx(index)]

    def isFull(self):
        return self.size == self.capacity

    def swap(self, idx_1, idx_2):
        temp = self.storage[idx_1]
        self.storage[idx_1] = self.storage[idx_2]
        self.storage[idx_2] = temp

    def insert(self, data):
        if(self.isFull()):
            raise("Heap is full")
        self.storage[self.size] = data;
        self.size += 1
        self.heapifyUp()

    def heapifyUp(self):
        index = self.size - 1
        while self.hasParent(index) and self.getParent(index) > self.storage[index]:
            self.swap(self.getParentIdx(index), index)
            index = self.getParentIdx(index)                                            


    def removeMin(self):
        if self.size == 0:
            raise("Empty heap")

        data = self.storage[0]
        self.storage[0] = self.storage[self.size-1]
        self.storage[self.size - 1] = 0
        self.size -= 1    
        self.heapifyDown()
        return data

    def heapifyDown(self):
        index = 0

        while self.hasLeftChild(index):
            smallerChildIndex = self.getLeftChildIdx(index)
            if self.hasRightChild(index) and self.getRightChild(index) < self.getLeftChild(index):
                smallerChildIndex = self.getRightChild(index)

            if self.storage[index] < self.storage[smallerChildIndex]:
                break
            else:
                self.swap(index, smallerChildIndex)   

            index = smallerChildIndex


heap = MinHeap(5)

heap.insert(15)
heap.insert(150)
heap.insert(10)

print(heap.removeMin())

print(heap.storage)

File: Heap/src/com/company/MinHeap.java
package com.company;

import java.util.ArrayList;
import java.util.List;

public class MinHeap {
    
    List<Integer> storage;
    public MinHeap(){
        storage = new ArrayList<>();
    }
    
    boolean isEmpty(){ return storage.size() <= 0; }
    
    private int getParentIdx(int idx){ return (idx - 1) / 2; }
    private int getLeftChildIdx(int idx){ return 2 * idx + 1; }
    private int getRightChildIdx(int idx){ return 2 * idx + 2; }
    
    private boolean hasParent(int idx){ return getParentIdx(idx) >= 0; }
    private boolean hasLeftChild(int idx){ return getLeftChildIdx(idx) < storage.size(); }
    private boolean hasRightChild(int idx){ return getRightChildIdx(idx) < storage.size(); }
    
    private int getParentElement(int idx){ return storage.get(getParentIdx(idx)); }
    private int getLeftChildElement(int idx){ return storage.get(this.getLeftChildIdx(idx)); }
    private int getRightChildElement(int idx){ return storage.get(this.getRightChildIdx(idx)); }
    
    void swap(int i, int j){
        
        int temp = storage.get(i);
        storage.set(i, storage.get(j));
        storage.set(j, temp);        
        
    }
    
    private void heapifyUp(){
        
        int idx = storage.size() - 1;
        while(hasParent(idx) && getParentElement(idx) > storage.get(idx)){
            swap(getParentIdx(idx), idx);
            idx = getParentIdx(idx);
        }
        
    }
    
    private void heapifyDown(){
        
        int idx = 0;
        while(hasLeftChild(idx)){
            
            int smallerChildIdx = getLeftChildIdx(idx);
            if(hasRightChild(idx) && getRightChildElement(idx) < getLeftChildElement(idx))
                smallerChildIdx = getRightChildIdx(idx);
            
            if(storage.get(idx) < storage.get(smallerChildIdx))
                break;
            else
                swap(idx, smallerChildIdx);
            
            idx = smallerChildIdx;
            
        }
        
    }
    
    void insert(int data){        
        storage.add(data);
        heapifyUp();        
    }
    
    int removeMin(){
        
        if(isEmpty()) return Integer.MIN_VALUE;
        
        int data = storage.remove(0);
        
        if(storage.size() > 0){

            int temp = storage.remove(storage.size() - 1);

            storage.add(0, temp);

            heapifyDown();
            
        }
        
        return data;
        
    }
    
    
}
